File: appium/utils/driver_factory.py
"""
Driver factory for Appium E2E.

- E2E_DRIVER=appium  -> real Appium UiAutomator2 session
- E2E_DRIVER=simulated -> deterministic flow driver for CI / offline runs
"""

from __future__ import annotations

from dataclasses import dataclass, field

from config import settings


class ElementNotFound(Exception):
    pass


@dataclass
class SimElement:
    text: str = ""
    content_desc: str = ""
    enabled: bool = True

    def click(self) -> None:
        return None

    def clear(self) -> None:
        return None

    def send_keys(self, value: str) -> None:
        self.text = value


@dataclass
class SimulatedDriver:
    """In-memory UI state machine mirroring Smart Furnace Android screens."""

    screen: str = "login"
    fields: dict[str, str] = field(default_factory=dict)
    visible_texts: set[str] = field(default_factory=set)
    logged_in: bool = False
    feasibility_message: str = ""

    def __post_init__(self) -> None:
        self._show_login()

    def quit(self) -> None:
        return None

    def implicit_wait(self, _seconds: int) -> None:
        return None

    def _show_login(self) -> None:
        self.screen = "login"
        self.visible_texts = {
            "SMART FURNACE AI",
            "Advanced Material Synthesis",
            "Scientist Authorization",
            "Institutional Email",
            "Password",
            "ACCESS TERMINAL",
            "No access? Request an account here.",
        }

    def _show_register(self) -> None:
        self.screen = "register"
        self.visible_texts = {
            "REQUEST ACCESS",
            "Full Name",
            "Institutional Email",
            "Create Password",
            "REGISTER SCIENTIST",
            "Already authorized? Access Terminal here.",
        }

    def _show_home(self) -> None:
        self.screen = "home"
        self.logged_in = True
        self.visible_texts = {
            "Home",
            "Dashboard",
            "History",
            "Settings",
            "Welcome to Smart Furnace AI",
            "About Our Product",
            "Key Features",
            "How It Works",
            "Advanced Technology",
        }

    def _show_dashboard(self) -> None:
        self.screen = "dashboard"
        self.visible_texts = {
            "Home",
            "Dashboard",
            "History",
            "Settings",
            "Synthesis Dashboard",
            "Synthesis Parameters",
            "Standard Synthesis",
            "Phase-Specific",
            "Base Material",
            "Target Material",
            "CHECK FEASIBILITY",
        }

    def _show_history(self) -> None:
        self.screen = "history"
        self.visible_texts = {
            "Home",
            "Dashboard",
            "History",
            "Settings",
            "Synthesis History",
            "Experiment Records",
            "Successful",
            "Total Experiments",
            "Success Rate",
            "No experiment history found for this user yet.",
            "Refresh",
        }

    def _show_settings(self) -> None:
        self.screen = "settings"
        self.visible_texts = {
            "Home",
            "Dashboard",
            "History",
            "Settings",
            "User Profile",
            "Preferences",
            "Account",
            "Notifications",
            "Dark Mode",
            "LOGOUT",
            "Confirm Logout",
            "Cancel",
            "Logout",
        }

    def find_element(self, by: str, value: str) -> SimElement:
        key = value
        mapping = {
            "login_email_field": "email",
            "login_password_field": "password",
            "login_submit_button": "ACCESS TERMINAL",
            "login_register_link": "No access? Request an account here.",
            "register_name_field": "name",
            "register_email_field": "email",
            "register_password_field": "password",
            "register_submit_button": "REGISTER SCIENTIST",
            "dashboard_base_material_field": "base_material",
            "dashboard_target_material_field": "target_material",
            "dashboard_check_feasibility_button": "CHECK FEASIBILITY",
            "settings_logout_button": "LOGOUT",
        }

        if by in {"id", "accessibility id", "-android uiautomator"}:
            if key in mapping and mapping[key] in {
                "email",
                "password",
                "name",
                "base_material",
                "target_material",
            }:
                return SimElement(text=self.fields.get(mapping[key], ""))
            if key in mapping:
                return SimElement(text=mapping[key])

        if by == "xpath" and "ACCESS TERMINAL" in value:
            return SimElement(text="ACCESS TERMINAL")
        if by == "xpath" and "REGISTER SCIENTIST" in value:
            return SimElement(text="REGISTER SCIENTIST")
        if by == "xpath" and "CHECK FEASIBILITY" in value:
            return SimElement(text="CHECK FEASIBILITY")
        if by == "xpath" and "LOGOUT" in value:
            return SimElement(text="LOGOUT")
        if by == "xpath" and "Logout" in value:
            return SimElement(text="Logout")

        raise ElementNotFound(f"{by}={value}")

    def find_elements(self, by: str, value: str) -> list[SimElement]:
        try:
            return [self.find_element(by, value)]
        except ElementNotFound:
            return []

    def find_by_text(self, text: str) -> SimElement:
        if text not in self.visible_texts and text not in {
            "ACCESS TERMINAL",
            "REGISTER SCIENTIST",
            "CHECK FEASIBILITY",
            "LOGOUT",
            "Logout",
            "Cancel",
            "Home",
            "Dashboard",
            "History",
            "Settings",
        }:
            # Allow navigation labels always while logged in
            if self.logged_in and text in {"Home", "Dashboard", "History", "Settings"}:
                return SimElement(text=text)
            raise ElementNotFound(text)
        return SimElement(text=text)

    def tap_text(self, text: str) -> None:
        if text == "No access? Request an account here.":
            self._show_register()
            return
        if text == "Already authorized? Access Terminal here.":
            self._show_login()
            return
        if text == "ACCESS TERMINAL":
            email = self.fields.get("email", "")
            password = self.fields.get("password", "")
            if "@" in email and len(password) > 0:
                self._show_home()
            return
        if text == "REGISTER SCIENTIST":
            name = self.fields.get("name", "")
            email = self.fields.get("email", "")
            password = self.fields.get("password", "")
            if len(name) >= 2 and "@" in email and len(password) >= 6:
                self._show_home()
            return
        if text == "Home":
            self._show_home()
            return
        if text == "Dashboard":
            self._show_dashboard()
            return
        if text == "History":
            self._show_history()
            return
        if text == "Settings":
            self._show_settings()
            return
        if text == "CHECK FEASIBILITY":
            base = self.fields.get("base_material", "").strip()
            target = self.fields.get("target_material", "").strip()
            if not base or not target:
                self.feasibility_message = "Base and Target materials are required."
                self.visible_texts.add(self.feasibility_message)
            else:
                self.feasibility_message = "Feasibility review"
                self.visible_texts.add("Feasibility review")
                self.visible_texts.add("Message")
                self.visible_texts.add("Reason")
            return
        if text == "LOGOUT":
            self.visible_texts.add("Confirm Logout")
            return
        if text == "Logout":
            self.logged_in = False
            self.fields.clear()
            self._show_login()
            return
        if text == "Cancel":
            self.visible_texts.discard("Confirm Logout")
            return

    def type_into(self, field_key: str, value: str) -> None:
        self.fields[field_key] = value

    def page_source_contains(self, text: str) -> bool:
        return text in self.visible_texts or text == self.feasibility_message


def create_driver():
    if settings.E2E_DRIVER == "appium":
        from appium import webdriver
        from appium.options.android import UiAutomator2Options

        options = UiAutomator2Options()
        options.platform_name = "Android"
        options.device_name = "Android Emulator"
        options.app = settings.APK_PATH
        options.app_package = settings.APP_PACKAGE
        options.app_activity = settings.APP_ACTIVITY
        options.automation_name = "UiAutomator2"
        options.no_reset = False
        options.new_command_timeout = 120
        driver = webdriver.Remote(settings.APPIUM_SERVER_URL, options=options)
        driver.implicitly_wait(settings.IMPLICIT_WAIT_SEC)
        return driver

    return SimulatedDriver()

File: selenium/utils/driverFactory.js
/**
 * Driver factory for Selenium E2E.
 * - E2E_DRIVER=selenium  -> real Chrome WebDriver
 * - E2E_DRIVER=simulated -> deterministic web flow driver for CI
 */

const { Builder, By, until } = require("selenium-webdriver");
const chrome = require("selenium-webdriver/chrome");
const settings = require("../config/settings");

class SimulatedDriver {
  constructor() {
    this.screen = "login";
    this.fields = {};
    this.visibleTexts = new Set();
    this.loggedIn = false;
    this.feasibilityMessage = "";
    this._showLogin();
  }

  async quit() {}

  _showLogin() {
    this.screen = "login";
    this.visibleTexts = new Set([
      "SMART FURNACE AI",
      "Advanced Material Synthesis",
      "Scientist Authorization",
      "Institutional Email",
      "Password",
      "ACCESS TERMINAL",
      "No access? Request an account here.",
    ]);
  }

  _showRegister() {
    this.screen = "register";
    this.visibleTexts = new Set([
      "REQUEST ACCESS",
      "Full Name",
      "Institutional Email",
      "Create Password",
      "REGISTER SCIENTIST",
      "Already authorized? Access Terminal here.",
    ]);
  }

  _showHome() {
    this.screen = "home";
    this.loggedIn = true;
    this.visibleTexts = new Set([
      "Home",
      "Dashboard",
      "History",
      "Settings",
      "Welcome to Smart Furnace AI",
      "About Our Product",
      "Key Features",
      "How It Works",
      "Advanced Technology",
    ]);
  }

  _showDashboard() {
    this.screen = "dashboard";
    this.visibleTexts = new Set([
      "Home",
      "Dashboard",
      "History",
      "Settings",
      "Synthesis Dashboard",
      "Synthesis Parameters",
      "Standard Synthesis",
      "Phase-Specific",
      "Base Material",
      "Target Material",
      "CHECK FEASIBILITY",
    ]);
  }

  _showHistory() {
    this.screen = "history";
    this.visibleTexts = new Set([
      "Home",
      "Dashboard",
      "History",
      "Settings",
      "Synthesis History",
      "Experiment Records",
      "Successful",
      "Total Experiments",
      "Success Rate",
      "No experiment history found for this user yet.",
      "Refresh",
    ]);
  }

  _showSettings() {
    this.screen = "settings";
    this.visibleTexts = new Set([
      "Home",
      "Dashboard",
      "History",
      "Settings",
      "User Profile",
      "Preferences",
      "Account",
      "Notifications",
      "Dark Mode",
      "LOGOUT",
      "Confirm Logout",
      "Cancel",
      "Logout",
    ]);
  }

  async get(url) {
    this._showLogin();
    this.currentUrl = url;
  }

  async findByText(text) {
    if (
      !this.visibleTexts.has(text) &&
      !(this.loggedIn && ["Home", "Dashboard", "History", "Settings"].includes(text))
    ) {
      throw new Error(`Text not found: ${text}`);
    }
    return {
      click: async () => this.tapText(text),
      clear: async () => {},
      sendKeys: async (value) => {
        // unused for text buttons
        this.lastTyped = value;
      },
      getText: async () => text,
    };
  }

  async typeInto(fieldKey, value) {
    this.fields[fieldKey] = value;
  }

  async tapText(text) {
    if (text === "No access? Request an account here.") {
      this._showRegister();
      return;
    }
    if (text === "Already authorized? Access Terminal here.") {
      this._showLogin();
      return;
    }
    if (text === "ACCESS TERMINAL") {
      const email = this.fields.email || "";
      const password = this.fields.password || "";
      if (email.includes("@") && password.length > 0) this._showHome();
      return;
    }
    if (text === "REGISTER SCIENTIST") {
      const name = this.fields.name || "";
      const email = this.fields.email || "";
      const password = this.fields.password || "";
      if (name.length >= 2 && email.includes("@") && password.length >= 6) {
        this._showHome();
      }
      return;
    }
    if (text === "Home") return this._showHome();
    if (text === "Dashboard") return this._showDashboard();
    if (text === "History") return this._showHistory();
    if (text === "Settings") return this._showSettings();
    if (text === "CHECK FEASIBILITY") {
      const base = (this.fields.base_material || "").trim();
      const target = (this.fields.target_material || "").trim();
      if (!base || !target) {
        this.feasibilityMessage = "Base and Target materials are required.";
        this.visibleTexts.add(this.feasibilityMessage);
      } else {
        this.feasibilityMessage = "Feasibility review";
        this.visibleTexts.add("Feasibility review");
        this.visibleTexts.add("Message");
        this.visibleTexts.add("Reason");
      }
      return;
    }
    if (text === "LOGOUT") {
      this.visibleTexts.add("Confirm Logout");
      return;
    }
    if (text === "Logout") {
      this.loggedIn = false;
      this.fields = {};
      this._showLogin();
      return;
    }
    if (text === "Cancel") {
      this.visibleTexts.delete("Confirm Logout");
    }
  }

  async pageContains(text) {
    return this.visibleTexts.has(text) || text === this.feasibilityMessage;
  }
}

class SeleniumDriverAdapter {
  constructor(driver) {
    this.driver = driver;
  }

  async quit() {
    await this.driver.quit();
  }

  async get(url) {
    await this.driver.get(url);
  }

  async findByText(text) {
    const locator = By.xpath(`//*[contains(normalize-space(.), "${text}")]`);
    const el = await this.driver.wait(until.elementLocated(locator), settings.implicitWaitMs);
    return el;
  }

  async typeInto(fieldKey, value) {
    // Prefer Flutter semantics / aria labels exposed on web.
    const labelMap = {
      email: "Institutional Email",
      password: "Password",
      name: "Full Name",
      base_material: "Base Material",
      target_material: "Target Material",
    };
    const label = labelMap[fieldKey] || fieldKey;
    const input = await this.driver.findElement(
      By.xpath(
        `//*[contains(normalize-space(.), "${label}")]/following::input[1] | //*[@aria-label="${label}"] | //flt-semantics[@aria-label="${label}"]`
      )
    );
    await input.clear();
    await input.sendKeys(value);
  }

  async tapText(text) {
    const el = await this.findByText(text);
    await el.click();
  }

  async pageContains(text) {
    try {
      await this.findByText(text);
      return true;
    } catch {
      const source = await this.driver.getPageSource();
      return source.includes(text);
    }
  }
}

async function createDriver(mode = settings.driverMode) {
  if (mode === "selenium") {
    const options = new chrome.Options();
    if (settings.headless) {
      options.addArguments("--headless=new", "--window-size=1280,900");
    }
    options.addArguments("--disable-gpu", "--no-sandbox", "--disable-dev-shm-usage");
    const driver = await new Builder().forBrowser("chrome").setChromeOptions(options).build();
    await driver.manage().setTimeouts({ implicit: settings.implicitWaitMs });
    return new SeleniumDriverAdapter(driver);
  }
  return new SimulatedDriver();
}

module.exports = { createDriver, SimulatedDriver, SeleniumDriverAdapter };
